### PR TITLE
[XLA:GPU] Enhance cuDNN fused attention rewriter patttern match with num of users check

### PR DESCRIPTION
**Pull request by Cjkkkk**
Follow up PR for https://github.com/openxla/xla/pull/7016.

Enhance current fused attention rewriter pattern match by adding more checks for num of users of matched nodes. For example, for softmax sub pattern like `sub(x, reduce_max(x))`, we want to make sure `sub` and `reduce_max` node has exactly one user, therefore no softmax nodes will be left in the graph after softmax is replaced with cudnn fused attention custom call.

1. We add checks for all functional nodes that we pattern match in the rewriter. For example, in pattern bmm1-bias-softmax-bmm2, for softmax, it would be sub/reduce/divide/exp nodes and for bias add, it would be add node. For each of these nodes, it should either have exactly one user or less than or equal two users if it could be used by backwards.
2. Remove some back to back optional convert.
3. Add some functions in pattern match to easily check num of users(`WithNumUser()` and `WithAtMostNumUser()`)

### CODE REVIEW

**Comment by Cjkkkk** [comment]
@akuegel Hi Adrian, this is the follow up PR for https://github.com/openxla/xla/pull/7016 as we discussed last week.

**Comment by Cjkkkk** [comment]
is this test `Int8GemmRankGreaterThanTwo` expected to fail? @akuegel

**Comment by Cjkkkk** [comment]
seems like file check issue. Fixed with a generic name.

**Comment by Cjkkkk** [comment]
Not sure why https://github.com/openxla/xla/commit/0e01a692222c041f2703e6c113e92991b1e80d61 changes this test file check.

**Comment by akuegel** [comment]
> Not sure why [0e01a69](https://github.com/openxla/xla/commit/0e01a692222c041f2703e6c113e92991b1e80d61) changes this test file check.

When enabling the FMHA pipeline, it effectively means that we will do Layout normalization before running GemmRewriterTriton, and before it was done after GemmRewriterTriton. This can affect test expectations in this test, as those are End2End tests.

File: xla/service/pattern_matcher.h [hunk 1]
```diff
@@ -1867,6 +1867,55 @@ class HloInstructionPatternOneUserImpl
...
+      EXPLAIN << "HloInstruction has " << inst->user_count()
+              << " users, but expected less than or equal "
+              << user_num_ << " users.";
+      return false;
+    }
+    return true;
+  }
+
+  void DescribeTo(std::ostream* os, int64_t indent = 0) const {
+    *os << "which has less than or euqal " << user_num_
```
Comment (akuegel) [inline xla/service/pattern_matcher.h:1910]: nit: euqal -> equal

File: xla/service/pattern_matcher.h [hunk 2]
```diff
@@ -1867,6 +1867,55 @@ class HloInstructionPatternOneUserImpl
...
+           "same instruction)";
+  }
+
+  private:
+  int64_t user_num_;
+};
+
+class HloInstructionPatternLeNumUserImpl {
+ public:
+  explicit constexpr HloInstructionPatternLeNumUserImpl(
```
Comment (akuegel) [inline xla/service/pattern_matcher.h:1896]: nit: maybe rename and use AtMost instead of Le ?

File: xla/service/pattern_matcher.h [hunk 3]
```diff
@@ -2418,6 +2467,18 @@ class HloInstructionPattern {
...
 
+  // Modifies the pattern to match if the instruction is used by exactly user_num
+  // times by other instruction.
+  constexpr auto WithNumUser(int64_t user_num) const {
+    return AppendImpl(HloInstructionPatternNumUserImpl(user_num));
+  }
+
+  // Modifies the pattern to match if the instruction is used by less than user_num
+  // times by other instruction.
+  constexpr auto WithLeNumUser(int64_t user_num) const {
```
Comment (akuegel) [inline xla/service/pattern_matcher.h:2478]: AtMost instead of Le?

**Review by akuegel — CHANGES_REQUESTED** [review]
I am fine with enabling the flag in the same PR, but I just want to warn that in case a rollback is needed, it could happen that the whole change is rolled back instead of just the flag flip. This is possibly outside my control, as the rollback request would most likely happen (and be approved) outside my working hours.

**Comment by AyanmoI** [comment]
> I am fine with enabling the flag in the same PR, but I just want to warn that in case a rollback is needed, it could happen that the whole change is rolled back instead of just the flag flip. This is possibly outside my control, as the rollback request would most likely happen (and be approved) outside my working hours.

Sure let's do that and create a separate PR.

**Comment by Cjkkkk** [comment]
> I am fine with enabling the flag in the same PR, but I just want to warn that in case a rollback is needed, it could happen that the whole change is rolled back instead of just the flag flip. This is possibly outside my control, as the rollback request would most likely happen (and be approved) outside my working hours.

I will move enabling the flag into a separate PR.

**Review by akuegel — APPROVED** [review]

**Comment by ezhulenev** [comment]
Fails in internal CI with: 

```
[third_party/tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc:3334](https://cs.corp.google.com/piper///depot/google3/third_party/tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc?l=3334&ws=tap-presubmit-server/133418086&snapshot=2):25: error: unused variable 'fmha' [-Werror,-Wunused-variable]
 3334 |   const HloInstruction* fmha;
      |                         ^~~~
[third_party/tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc:3384](https://cs.corp.google.com/piper///depot/google3/third_party/tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc?l=3384&ws=tap-presubmit-server/133418086&snapshot=2):25: error: unused variable 'fmha' [-Werror,-Wunused-variable]
 3384 |   const HloInstruction* fmha;
      |                         ^~~~
[third_party/tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc:3473](https://cs.corp.google.com/piper///depot/google3/third_party/tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc?l=3473&ws=tap-presubmit-server/133418086&snapshot=2):25: error: unused variable 'fmha' [-Werror,-Wunused-variable]
 3473 |   const HloInstruction* fmha;
      |                         ^~~~
3 errors generated.
```

**Comment by Cjkkkk** [comment]
> Fails in internal CI with:
> 
> ```
> [third_party/tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc:3334](https://cs.corp.google.com/piper///depot/google3/third_party/tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc?l=3334&ws=tap-presubmit-server/133418086&snapshot=2):25: error: unused variable 'fmha' [-Werror,-Wunused-variable]
>  3334 |   const HloInstruction* fmha;
>       |                         ^~~~
> [third_party/tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc:3384](https://cs.corp.google.com/piper///depot/google3/third_party/tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc?l=3384&ws=tap-presubmit-server/133418086&snapshot=2):25: error: unused variable 'fmha' [-Werror,-Wunused-variable]
>  3384 |   const HloInstruction* fmha;
>       |                         ^~~~
> [third_party/tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc:3473](https://cs.corp.google.com/piper///depot/google3/third_party/tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc?l=3473&ws=tap-presubmit-server/133418086&snapshot=2):25: error: unused variable 'fmha' [-Werror,-Wunused-variable]
>  3473 |   const HloInstruction* fmha;
>       |                         ^~~~
> 3 errors generated.
> ```

should be fixed now.

**Review by akuegel — APPROVED** [review]

**Comment by Cjkkkk** [comment]
seems still some internal checks failed. Could anyone shares some info on this?

**Pull request closed without merging**